### PR TITLE
chore: Add internal shareable ESLint config

### PR DESCRIPTION
**Pull request by chaance**
This PR introduces an internal ESLint config we can share across our org's packages. This config is not meant for public use, nor is it considered public API.

It also adds a new rule per `REM-784` and fixes issues in the Deno templates that would be flagged by the updated rules. Deno.

All other changes are a result of running `eslint --fix` after updating the config.

### CODE REVIEW

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]
Awesome work to make the internal ESLint rules a bit more defined. 💪👊

I was also already doing this in #2226, #2227 #2228 & #2229, so I added the changes I did there as a remark on this PR, together with some other remarks/questions I had.

File: .eslintrc.js
Comment (mjackson) [reply .eslintrc.js:35]: Yeah, we should definitely eliminate all node-specific code in remix-server-runtime. But not in remix-node, obviously. Is that not what this does?

File: docs/api/remix.md
Comment (mjackson) [inline docs/api/remix.md:101]: This should be a `let`. Let's remove `prefer-const` from the docs dir and just use one consistent style throughout the whole repo.

**Review by mjackson — CHANGES_REQUESTED** [review]
Thanks for taking this on, @chaance. I'm going to reneg on what I said last week and suggest we use `let` for anything that is an "intermediate" value. In other words, module-scope variables should probably be `const`. Everything else should be `let`, including in our docs.

File: .eslintrc.js
Comment (MichaelDeBoey) [reply .eslintrc.js:35]: This rule is targeting the following files:

```ts
      files: [
        "packages/remix-server-runtime/**/*.js?(x)",
        "packages/remix-server-runtime/**/*.ts?(x)",
      ],
```

File: packages/remix-eslint-config/internal.js
Comment (chaance) [reply packages/remix-eslint-config/internal.js:20]: I'm keeping the markdown rules/plugin out of the shared config for now because we'll be using this in outside Remix repos that won't use them. Might move them in later.

File: packages/remix-eslint-config/internal.js
Comment (chaance) [reply packages/remix-eslint-config/internal.js:61]: For now I think we'll leave this as-is. Reason is that we'll use this config outside of the Remix repo (React Router repo, for example) that may include JSX/React in non-jsx file extensions.

File: .eslintrc.js
Comment (chaance) [reply .eslintrc.js]: I updated this and just added an override config in the `remix-server-runtime` project for now. I also added this override to the Deno adapter.

File: docs/api/remix.md
Comment (chaance) [reply docs/api/remix.md]: This should be reverted. Reminder that we'll handle all the stuff in the docs ahead of the next release when we merge everything from `main`.

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [reply packages/remix-eslint-config/internal.js:61]: This is only about MD files though, not about the actual code

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:77]: I don't think adding rules to the docs folder would help, as they only contain MD files, which are already covered above
```suggestion
```

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:91]: > I don't think adding rules to the docs folder would help, as they only contain MD files, which are already covered above

```suggestion
```

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:83]: Do we really want to enable unused expressions/labels/vars in examples?  
These will only clutter the examples unnecessary imo.

```suggestion
```

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:95]: > Do we really want to enable unused expressions/labels/vars in examples?  
> These will only clutter the examples unnecessary imo.

```suggestion
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: packages/create-remix/templates/.eslintrc.js
Comment (mjackson) [inline packages/create-remix/templates/.eslintrc.js:3]: Why not?

File: package.json
Comment (mjackson) [inline package.json:16]: Why did you remove the `--fix`? Is this the part you were telling me about earlier where you said it won't be active yet?

File: packages/remix-deno/globals.ts
Comment (mjackson) [inline packages/remix-deno/globals.ts:13]: Seems like we should open an issue on the `prefer-let` code to explain this use case to them and ask them for the ability to opt-out in global TS declarations?

File: packages/remix-eslint-config/internal.js
Comment (mjackson) [inline packages/remix-eslint-config/internal.js:89]:
```suggestion
```

**Review by mjackson — CHANGES_REQUESTED** [review]
Looks pretty good, but I was confused about why this PR creates files in the deno template. Shouldn't those be created in a separate PR?

File: package.json
Comment (chaance) [reply package.json:16]: Ah, oversight. I removed it temporarily for dev purposes. Adding back.

File: packages/create-remix/templates/.eslintrc.js
Comment (MichaelDeBoey) [reply packages/create-remix/templates/.eslintrc.js:3]: It's user-facing code, so I think we _should_ care tbh.

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:75]: ff9db8210390b0067510a9c428d49a2b0cfe5f08

> I don't think adding rules to the docs folder would help, as they only contain MD files, which are already covered above
> 
> ```suggestion
> ```

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:90]: ff9db8210390b0067510a9c428d49a2b0cfe5f08

> > I don't think adding rules to the docs folder would help, as they only contain MD files, which are already covered above
> 
> ```suggestion
> ```

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:95]: ff9db8210390b0067510a9c428d49a2b0cfe5f08

> > Do we really want to enable unused expressions/labels/vars in examples?  
> > These will only clutter the examples unnecessary imo.
> 
> ```suggestion
> ```

File: packages/remix-eslint-config/internal.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/internal.js:97]: d6fbe5a3f2fbfa63d75b80c52232ea38c8600e22

```suggestion
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]
@chaance It seems like you accidentally reverted ff9db8210390b0067510a9c428d49a2b0cfe5f08 & d6fbe5a3f2fbfa63d75b80c52232ea38c8600e22 when merging branches

File: packages/remix-eslint-config/internal.js
Comment (chaance) [reply packages/remix-eslint-config/internal.js:83]: I think so. Examples may show how an object is initialized or where it's imported from without necessarily using every single reference. We'll just use our best judgment here on a case-by-case basis.

File: docs/api/remix.md
Comment (MichaelDeBoey) [reply docs/api/remix.md]: @mjackson Does this mean we don't want any user-facing code (so templates + docs) to use `const` anymore (like suggested/introduced by @kentcdodds a couple of weeks/months(?) ago) and _always_ use `let`?

File: package.json
Comment (MichaelDeBoey) [inline package.json:78]: Shouldn't this be a dependency of `@remix-run/eslint-config` as we're now having extending `@remix-run/eslint-config/node` in the configs?

File: packages/remix-eslint-config/node.js
Comment (MichaelDeBoey) [inline packages/remix-eslint-config/node.js:8]: I think it doesn't make sense to have `eslint-plugin-node` in the `plugins` array if we're enabling any rules.

So I think we should either enable some basic rules **or** we should remove it from the `plugins` array

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: package.json
Comment (chaance) [reply package.json:78]: It already is. https://github.com/remix-run/remix/blob/main/packages/remix-eslint-config/package.json#L28

**Review by mjackson — COMMENTED** [review]
This all looks pretty good to me but I'd like to hold off on merging until we can get Logan and Pedro's stuff merged from this cycle. Hopefully we can merge this on Monday after re-running it over the entire repo.

**Pull request closed without merging**
